spring-cloud-azure-config/src/main/java/com/microsoft/azure/spring/cloud/config/AzureCloudConfigWatch.java

```diff
@@ -0,0 +1,136 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.cloud.endpoint.event.RefreshEvent;
+import org.springframework.context.ApplicationEventPublisher;
+import org.springframework.context.ApplicationEventPublisherAware;
+import org.springframework.context.SmartLifecycle;
+import org.springframework.scheduling.TaskScheduler;
+import org.springframework.util.StringUtils;
+
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Optional;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.stream.Collectors;
+
+public class AzureCloudConfigWatch implements ApplicationEventPublisherAware, SmartLifecycle {
+    private static final Logger LOGGER = LoggerFactory.getLogger(AzureCloudConfigWatch.class);
+    private final ConfigServiceOperations configOperations;
+    private final LinkedHashMap<String, String> keyNameEtagMap = new LinkedHashMap<>();
+    private final TaskScheduler taskScheduler;
+    private final AtomicBoolean running = new AtomicBoolean(false);
+    private ApplicationEventPublisher publisher;
+    private ScheduledFuture<?> watchFuture;
+    private AzureCloudConfigProperties properties;
+    private boolean firstTime = true;
+
+    public AzureCloudConfigWatch(ConfigServiceOperations operations, AzureCloudConfigProperties properties,
+                                 TaskScheduler scheduler) {
+        this.configOperations = operations;
+        this.properties = properties;
+        this.taskScheduler = scheduler;
+    }
+
+    @Override
+    public void setApplicationEventPublisher(ApplicationEventPublisher applicationEventPublisher) {
+        this.publisher = applicationEventPublisher;
+    }
+
+    @Override
+    public boolean isAutoStartup() {
+        return true;
+    }
+
+    @Override
+    public void stop(Runnable callback) {
+        this.stop();
+        callback.run();
+    }
+
+    @Override
+    public void start() {
+        if (this.running.compareAndSet(false, true)) {
+            this.watchFuture = this.taskScheduler.scheduleWithFixedDelay(this::watchConfigKeyValues,
+                    this.properties.getWatch().getDelay());
+        }
+    }
+
+    @Override
+    public void stop() {
+        if (this.running.compareAndSet(true, false) && this.watchFuture != null) {
+            this.watchFuture.cancel(true);
+        }
+    }
+
+    @Override
+    public boolean isRunning() {
+        return this.running.get();
+    }
+
+    @Override
+    public int getPhase() {
+        return 0;
+    }
+
+    public void watchConfigKeyValues() {
+        if (this.running.get()) {
+            String prefix = StringUtils.hasText(properties.getPrefix()) ? properties.getPrefix() + "*" : "*";
+            List<KeyValueItem> keyValueItems = configOperations.getKeys(prefix, properties.getLabel());
+
+            if (keyValueItems == null || keyValueItems.isEmpty()) {
```

Comment:
Will null a valid return value? If not, we should not handle it.

Reply:
Will not return null, removed null check.